2020/4/passport/src/main.rs

```diff
@@ -0,0 +1,248 @@
+    let file = File::open(&args[1]);
+
+    let reader = io::BufReader::new(file.unwrap());
+    let mut current = String::from("");
+    let mut passports = Vec::new();
+    for line in reader.lines() {
+        let l = line.unwrap();
+        if l.len() == 0 {
+            let passport = Passport::from_string(&current);
+            passports.push(passport);
+            current = String::from("");
+        } else {
+            current = current + " " + &l;
+        }
+    }
+    let passport = Passport::from_string(&current);
+    passports.push(passport);
+
+    println!("Found {} total passports", passports.len());
+    println!(
+        "Found {} valid passports",
+        passports.iter().filter(|p| p.is_valid()).count()
+    );
+}
+
+pub struct Passport {
+    birth_year: Option<String>,
+    issue_year: Option<String>,
+    expiration_year: Option<String>,
+    height: Option<String>,
+    hair_color: Option<String>,
+    eye_color: Option<String>,
+    passport_id: Option<String>,
+    country_id: Option<String>,
+}
+
+impl Passport {
+    pub fn from_string(data: &str) -> Passport {
+        let mut birth_year = None;
+        let mut issue_year = None;
+        let mut expiration_year = None;
+        let mut height = None;
+        let mut hair_color = None;
+        let mut eye_color = None;
+        let mut passport_id = None;
+        let mut country_id = None;
+        let parts = data.split_whitespace();
+        for p in parts {
+            let mut items = p.split(':');
+            let name = items.next().unwrap();
+            let value = items.next().unwrap();
+            match name {
+                "byr" => {
+                    birth_year = Some(String::from(value));
+                }
+                "iyr" => {
+                    issue_year = Some(String::from(value));
+                }
+                "eyr" => {
+                    expiration_year = Some(String::from(value));
+                }
+                "hgt" => {
+                    height = Some(String::from(value));
+                }
+                "hcl" => {
+                    hair_color = Some(String::from(value));
+                }
+                "ecl" => {
+                    eye_color = Some(String::from(value));
+                }
+                "pid" => {
+                    passport_id = Some(String::from(value));
+                }
+                "cid" => {
+                    country_id = Some(String::from(value));
+                }
+
+                &_ => {}
+            }
+        }
+        Passport {
+            birth_year: birth_year,
+            issue_year: issue_year,
+            expiration_year: expiration_year,
+            height: height,
+            hair_color: hair_color,
+            eye_color: eye_color,
+            passport_id: passport_id,
+            country_id: country_id,
+        }
+    }
+
+    pub fn is_valid(&self) -> bool {
+        match &self.birth_year {
```

Comment:
This is probably not what I'd do :). Nesting is rough in most languages, Rust included.

I'd probably just have `is_valid_xxx` accept an `Option<T>`. You're never going to call it with a raw `T` so it's reasonable. There are even ways to make it accept both but that would involve defining your own traits since Rust doesn't support parameter overloading (but you can make it look like it because you can create a custom Trait and implement it for all the parameter types you want to accept).

You *could* also do this if you don't want to accept `Option<T>` in your helpers:

```rust
self.birth_year.map(|y| is_valid_year(y, 1920, 2020).unwrap_or(false) && ...
```

Using `map` (which converts `Option<T>` to `Option<U>` using a func from `T -> U`) and `unwrap_or` (which converts `Option<T>` to `T` by setting a default value for `T` when the `Option` is `None`).

Reply:
This kind of `Option` chaining is where I find Rust gets a little... verbose. You actually _can_ use `?` to unwrap `Option`s and it will cause your function to return `NoneError` if the option ends up being `None`. So another way to do this would be:

```rust
// Only Ok(true) is valid, all other returns are invalid.
pub fn is_valid(&self) -> Result<bool, NoneError> {
    is_valid_year(self.birth_year?, 1920, 2002) &&
        is_valid_year(self.issue_year?, 2010, 2020) &&
        // ...
}
```

You could also do the ol' wrap it in a helper:

```rust
// Only Ok(true) is valid, all other returns are invalid.
fn is_valid_helper(&self) -> Result<bool, NoneError> {
    is_valid_year(self.birth_year?, 1920, 2002) &&
        is_valid_year(self.issue_year?, 2010, 2020) &&
        // ...
}

pub fn is_valid(&self) -> bool { self.is_valid_helper().ok_or(false) }
```

Reply:
Good to know - I wasn't thrilled with this at all.